### DISCUSSION TITLE
Linear Pattern

### DISCUSSION THREAD

**Discussion by jgomez720** [KittyCAD/modeling-app#256] [New Feature Requests]
### Description 
A tool that enables users to efficiently replicate and arrange selected design elements or features along a specified linear path.

### Functionality
1. User selects sketch entity that they want to pattern (circle, line, multiple lines, points, click and drag over multiple entities?)
2. User defines which axis they would like to pattern along (typically this is either an existing edge of the model or the absolute coordinate system). Users can reverse the direction with a button.
3. User defines how many repetitions they would like.
4. User defines the spacing between repetitions.

### Helpful Docs/Images

**OnShape Linear Pattern Feature (Sketching)**

![Screenshot 2023-11-06 at 1 10 03 PM](https://github.com/KittyCAD/modeling-app/assets/114548659/8e68738f-0eb2-4f70-8702-708bddf01773)
![Screenshot 2023-11-06 at 1 10 29 PM](https://github.com/KittyCAD/modeling-app/assets/114548659/4efb12ae-506f-4fee-9626-d243878c2444)
![Screenshot 2023-11-06 at 1 10 47 PM](https://github.com/KittyCAD/modeling-app/assets/114548659/d2bf37ed-09dd-4ac7-a7ae-2c3535abac35)
![Screenshot 2023-11-06 at 1 11 16 PM](https://github.com/KittyCAD/modeling-app/assets/114548659/9e14acb9-b9f1-4c15-a058-ec391b8e2dd6)

## PseudoKCL
This is how I would do the same thing as the example above in KCL
```
const box = startSketchOn('XY')
  |> startProfileAt([0, 0], %)
  |> line([0, 4], %)
  |> line([12, 0], %)
  |> line([0, -4], %)
  |> close(%)
  |> extrude(1, %)

const holeSketch = startSketchOn(endSketchFace: box) // My syntax to say "Sketch on the face at the end of the extrusion"
  |> circle([1.5, 3], 0.5)
  |> linearPattern({
    x: {
      // note: OnShape includes the original sketch as one of the repetitions, I suggest we don't do that. 
      // repetitions = 2 means there are 2 ADDITIONAL repeats after the original.
      repetitions: 2, 
      distance: 4.5,
    },
    y: {
      repetitions: 1,
      distance: 2,
    }
  })
```

**Comment by Irev-Dev** [comment]
@jgomez720 considering a 2d grid case would it make sense to:
1) Do nothing, if you want a grid, do two linear patterns back to back
2) Add a separate grid pattern feature
3) Roll in a grid feature into this feature, it's the same feature just with different input params to get a grid
?

**Comment by jgomez720** [comment]
@Irev-Dev my take would be to do **3**. Other programs use **Linear Pattern Feature** as the grid case as well. So the user can select one axis (OR two), and then set the spacing distance for each direction, as well as repetitions for both directions

**Comment by adamchalmers** [comment]
tiny nit:

if we name it "patternLinear" then users will type "pat" and get autocomplete for "patternLinear", "patternCircular", etc etc.

IMO this is a better autocomplete experience than typing "linea" and getting "linearPattern" because typing "pat" will show you all the available patterns, help you learn the API.

**Comment by jessfraz** [comment]
these are merged into main and KCL, there is a bug with holes that is being fixed in the engine but the API wont change!

**Comment by franknoirot** [comment]
I'm gonna reopen this because we need to discuss, design and implement the point-and-click flows.
